packages/@blockprotocol/graph/src/type-resolver/entity-type-to-typescript.ts

```diff
@@ -0,0 +1,216 @@
+  schema: EntityType,
+  depth: number,
+  resolvedUrisToType: UriToType,
+): Promise<CompiledType & {}> => {
+  if (resolvedUrisToType[schema.$id]) {
+    return resolvedUrisToType[schema.$id]!;
+  }
+
+  // if the cache is empty, this is the schema the external caller provided. we need to know to add boilerplate up top
+  const rootSchema = Object.keys(resolvedUrisToType).length === 0;
+
+  const typeName = generateTypeNameFromSchema(schema);
+
+  if (!typeName) {
+    throw new Error("Schema is missing a title");
+  }
+
+  // This first generated type is just going to cover the entity's properties – they have other fields typed elsewhere
+  const propertyTypeName = `${typeName}Properties`;
+
+  // here we import the types defined elsewhere which we rely on, e.g. Entity
+  let compiledSchema = rootSchema ? generateImportStatements() : "";
+
+  // Generate the type for FooProperties, representing the entity's 'own properties' (not links or linked entities)
+  compiledSchema += await compileSchema(
+    { ...schema, title: propertyTypeName },
+    {
+      additionalProperties: false, // @todo add additionalProperties: false to entity type JSON schemas
+      bannerComment: rootSchema ? bannerComment(schema.$id, depth) : "",
+    },
+  );
+
+  // This adds 'type Foo = Entity<FooProperties>;'
+  const entityTypeDefinition = generateEntityDefinition(
+    typeName,
+    propertyTypeName,
+  );
+  compiledSchema += entityTypeDefinition;
+
+  // using a type instead of an interface allows using FooProperties in Entity<FooProperties>
+  // matching group needed in case the compiler has made e.g. FooProperties1 & FooProperties2 to handle inheritance
+  // @todo we need to catch types generated internally by the library via inheritance (allOf), which we don't know about
+  //   either these need to be converted to types too, or we need a different approach to Entity<Properties>
+  compiledSchema = compiledSchema.replace(
+    new RegExp(`export interface (${typeName}Properties\\S*)`, "g"),
+    `export type $1 =`,
+  );
+
+  // temporary hack for the @todo mentioned above
+  // this covers the one type we know is currently generated via inheritance
+  compiledSchema = compiledSchema.replace(
+    /export interface Link/g,
+    "export type Link =",
+  );
+
+  const compiledType = { typeName, typeScriptString: compiledSchema };
+
+  // add the compiled type to our map of already-resolved types in case it's encountered again when following links
+  // eslint-disable-next-line no-param-reassign -- this is intentionally mutated when this function is called
+  resolvedUrisToType[schema.$id] = compiledType;
+
+  if (depth === 0) {
+    return compiledType;
+  }
+
+  // if we're following links, we want to generate various types that can be plugged into subgraph functions
+
+  // keep track of the possible destination types from this entity type for each link type it includes
+  const typeLinkMap: Record<string, string> = {};
```

Comment:
Can we type these things better? I'm guessing this could be 
```suggestion
  const typeLinkMap: Record<VersionedUri, string> = {};
```

Reply:
This and other similar comments are a result of me using `EntityType` generated from the metaschema, not the one from our types. I've switched to using ours and made the types stricter in https://github.com/blockprotocol/blockprotocol/pull/861/commits/19b584bd377bbcb6ac9dec81e11593544d92c4f7

We should expose `typedEntries` somewhere at some point, I've just copied it for now